File: backend/models/student_course.py
"""
backend/models/student_course.py
Strict Pydantic response models for student-facing course experience (Phase 4).
Excludes admin audit fields, creator IDs, draft statuses, and quiz answer keys.
"""

from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field, ConfigDict


class StudentCourseSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    slug: str
    short_description: Optional[str] = None
    description: Optional[str] = None
    thumbnail_url: Optional[str] = None
    category: Optional[str] = None
    difficulty: str
    estimated_duration_minutes: Optional[int] = None
    status: str
    published_at: Optional[str] = None
    modules_count: int = 0
    lessons_count: int = 0


class StudentCourseListResponse(BaseModel):
    total: int
    items: List[StudentCourseSummary]
    page: int
    page_size: int
    total_pages: int


class StudentLessonSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    module_id: str
    title: str
    slug: Optional[str] = None
    short_description: Optional[str] = None
    position: int
    estimated_duration_minutes: Optional[int] = None


class StudentQuizSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    module_id: str
    title: str
    description: Optional[str] = None


class StudentModuleSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    course_id: str
    title: str
    description: Optional[str] = None
    position: int
    lessons: List[StudentLessonSummary] = Field(default_factory=list)
    quiz: Optional[StudentQuizSummary] = None


class StudentCourseDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    slug: str
    short_description: Optional[str] = None
    description: Optional[str] = None
    thumbnail_url: Optional[str] = None
    category: Optional[str] = None
    difficulty: str
    estimated_duration_minutes: Optional[int] = None
    status: str
    published_at: Optional[str] = None
    modules_count: int = 0
    lessons_count: int = 0
    modules: List[StudentModuleSummary] = Field(default_factory=list)


class StudentLessonNavigationItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    module_id: str
    module_title: Optional[str] = None


class StudentLessonContent(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    schema_version: int = 1
    blocks: List[Dict[str, Any]] = Field(default_factory=list)


class StudentLessonDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    course: Dict[str, Any]
    module: Dict[str, Any]
    lesson: StudentLessonSummary
    content: StudentLessonContent
    prev_lesson: Optional[StudentLessonNavigationItem] = None
    next_lesson: Optional[StudentLessonNavigationItem] = None

File: frontend/types/student-course.ts
/**
 * frontend/types/student-course.ts
 * Strict TypeScript types for student course experience (Phase 4).
 * Strictly read-only; excludes admin authoring fields and quiz answer keys.
 */

import type { LessonBlock } from "./lesson-content";

export interface StudentCourseSummary {
  id: string;
  title: string;
  slug: string;
  short_description?: string | null;
  description?: string | null;
  thumbnail_url?: string | null;
  category?: string | null;
  difficulty: "beginner" | "intermediate" | "advanced" | string;
  estimated_duration_minutes?: number | null;
  status: string;
  published_at?: string | null;
  modules_count: number;
  lessons_count: number;
}

export interface StudentCourseListResponse {
  total: number;
  items: StudentCourseSummary[];
  page: number;
  page_size: number;
  total_pages: number;
}

export interface StudentLessonSummary {
  id: string;
  module_id: string;
  title: string;
  slug?: string | null;
  short_description?: string | null;
  position: number;
  estimated_duration_minutes?: number | null;
}

export interface StudentQuizSummary {
  id: string;
  module_id: string;
  title: string;
  description?: string | null;
}

export interface StudentModuleSummary {
  id: string;
  course_id: string;
  title: string;
  description?: string | null;
  position: number;
  lessons: StudentLessonSummary[];
  quiz?: StudentQuizSummary | null;
}

export interface StudentCourseDetail extends StudentCourseSummary {
  modules: StudentModuleSummary[];
}

export interface StudentLessonNavigationItem {
  id: string;
  title: string;
  module_id: string;
  module_title?: string | null;
}

export interface StudentLessonDetail {
  course: {
    id: string;
    title: string;
    slug: string;
  };
  module: {
    id: string;
    title: string;
    position: number;
  };
  lesson: StudentLessonSummary;
  content: {
    schema_version: number;
    blocks: LessonBlock[];
  };
  prev_lesson?: StudentLessonNavigationItem | null;
  next_lesson?: StudentLessonNavigationItem | null;
}
